Use payloadbuilder for build_block

SkandaBhat opened the pull request:
FIxes #12535

mattsse commented on crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:
```diff
@@ -214,234 +199,11 @@ pub trait LoadPendingBlock:
     /// block contract call using the parent beacon block root received from the CL.
     fn build_block(
         &self,
-        env: PendingBlockEnv,
+        _env: PendingBlockEnv,
     ) -> Result<(SealedBlockWithSenders, Vec<Receipt>), Self::Error>
     where
         EthApiError: From<ProviderError>,
     {
```
I'd rather not provide a default impl to prevent issues

mattsse commented on crates/rpc/rpc/src/eth/helpers/pending_block.rs:
```diff
@@ -20,14 +32,90 @@ where
+            None,                 // Optional best payload
+        );
+
+        // Instantiate the payload builder locally
+        let evm_config = self.evm_config().clone();
+        let payload_builder = EthereumPayloadBuilder::new(evm_config);
+
+        // Build the payload
+        match payload_builder.try_build(args).unwrap() {
+            BuildOutcome::Better { payload, .. } => {
+                let sealed_block = payload.block().clone();
```
can't this use the configured payload builder?

https://github.com/paradigmxyz/reth/blob/33730536f5e5de1403f2deb3c7dbe551879b9724/crates/rpc/rpc-eth-api/src/node.rs#L34-L35

mattsse commented on crates/rpc/rpc-builder/src/lib.rs:
```diff
@@ -2035,7 +2041,7 @@ impl TransportRpcModules {
     pub fn merge_http(&mut self, other: impl Into<Methods>) -> Result<bool, RegisterMethodError> {
         if let Some(ref mut http) = self.http {
-            return http.merge(other.into()).map(|_| true)
+            return http.merge(other.into()).map(|_| true);
```
```suggestion
            return http.merge(other.into()).map(|_| true)
```

mattsse commented on crates/rpc/rpc-builder/src/lib.rs:
```diff
@@ -2047,7 +2053,7 @@ impl TransportRpcModules {
     pub fn merge_ws(&mut self, other: impl Into<Methods>) -> Result<bool, RegisterMethodError> {
         if let Some(ref mut ws) = self.ws {
-            return ws.merge(other.into()).map(|_| true)
+            return ws.merge(other.into()).map(|_| true);
```
```suggestion
            return ws.merge(other.into()).map(|_| true)
```

mattsse commented on crates/rpc/rpc-builder/src/lib.rs:
```diff
@@ -2059,7 +2065,7 @@ impl TransportRpcModules {
     pub fn merge_ipc(&mut self, other: impl Into<Methods>) -> Result<bool, RegisterMethodError> {
         if let Some(ref mut ipc) = self.ipc {
-            return ipc.merge(other.into()).map(|_| true)
+            return ipc.merge(other.into()).map(|_| true);
```
```suggestion
            return ipc.merge(other.into()).map(|_| true)
```

mattsse commented on crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:
```diff
@@ -130,7 +115,7 @@ pub trait LoadPendingBlock:
                         .map_err(Self::Error::from_eth_err)?
                     {
-                        return Ok(Some((block, receipts)))
+                        return Ok(Some((block, receipts)));
```
```suggestion
                        return Ok(Some((block, receipts)))
```

mattsse commented on crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:
```diff
@@ -162,7 +147,7 @@ pub trait LoadPendingBlock:
                 Err(err) => {
                     debug!(target: "rpc", "Failed to build pending block: {:?}", err);
-                    return Ok(None)
+                    return Ok(None);
```
```suggestion
                    return Ok(None)
```

mattsse commented on crates/rpc/rpc/src/eth/core.rs:
```diff
@@ -151,7 +153,7 @@ where
     type Evm = EvmConfig;
     type Network = Network;
-    type PayloadBuilder = ();
+    type PayloadBuilder = EthereumPayloadBuilder<EvmConfig>;
```
we need to solve this a little bit differently, 
similar to how provider is integrated, via a generic. but for this pr this  should be okay.

mattsse reviewed (changes requested):
overall, this looks pretty good already

SkandaBhat commented:
@mattsse have addressed comments and created an issue https://github.com/paradigmxyz/reth/issues/12844 to track EthereumPayloadBuilder. This is ready for review.

mattsse reviewed (commented):
sorry this turned out to be more complex than anticipated and we had to solve this differntly
